### PR TITLE
C++: parse member variables using templates with forward declared types

### LINKED ISSUE #2867: C++ parser does not tag member variable if type is template with forward declared class
The name of the parser: C++

I wrote a [unit test](https://github.com/drichardson/ctags/tree/template-member-forward-decl/Units/parser-cxx.r/template-member-forward-declaration.d) that I believe should pass in [this branch](https://github.com/drichardson/ctags/tree/template-member-forward-decl):

```
$ python3 misc/units.py run Units/parser-cxx.r --units template-member-forward-declaration

Category: ROOT
------------------------------------------------------------
Testing template-member-forward-declaration as C++          failed (unexpected output)

Summary (see CMDLINE.tmp to reproduce without test harness)
------------------------------------------------------------
  #passed:                                0
  #FIXED:                                 0
  #FAILED (broken args.ctags?):           0
  #FAILED (unexpected-exit-status):       0
  #FAILED (unexpected-output):            1
        template-member-forward-declaration
  #skipped (features):                    0
  #skipped (languages):                   0
  #skipped (infinite-loop):               0
  #known-bugs:                            0
```

The content of input file:

```C++
template <class Class>
class A
{
public:
        A() { p = new Class(); }
        Class *p;
};

class B {
public:
        // forward declaration okay because only used as pointer in Template
        A<class C> a_c_forward;
};
```

Running ctags with the options `--sort=no --kinds-C++=m` produces tags output that is missing an entry for `a_c_forward`.

```
p       input.cpp       /^      Class *p;$/;"   m       class:A typeref:typename:Class *        file:
```

The tags output you expect:

```
p       input.cpp       /^      Class *p;$/;"   m       class:A typeref:typename:Class *        file:
a_c_forward     input.cpp       /^      A<class C> a_c_forward;$/;"     m       class:B typeref:typename:A<class C>     file:
```

The version of ctags:

```
Universal Ctags 5.9.0(3088bfc1), Copyright (C) 2015 Universal Ctags Team
Universal Ctags is derived from Exuberant Ctags.
Exuberant Ctags 5.8, Copyright (C) 1996-2009 Darren Hiebert
  Compiled: Feb 14 2021, 14:45:13
  URL: https://ctags.io/
  Optional compiled features: +wildcards, +regex, +iconv, +debug, +option-directory, +packcc
```

How do you get ctags binary: Built locally.

## Additional Info

Here is a ctags [trace](https://github.com/universal-ctags/ctags/files/5978793/trace.txt). You can see that the `class` token is what is causing the parser to not treat the reset of the line as a variable declaration.

### PR DESCRIPTION
**Pull request by drichardson**
Fixes  #2867.

### CODE REVIEW

**Comment by masatake** [comment]
Thank you. Could you squash the changes into one commit?
In addition, I would like you to use `C++: ...` as the prefix for the commit log.

**Comment by drichardson** [comment]
> 
> 
> Thank you. Could you squash the changes into one commit?
> In addition, I would like you to use `C++: ...` as the prefix for the commit log.

Done and done.

**Comment by masatake** [comment]
Thank you.
Let's wait for the maintainer to find the time for review:-)

**Review by masatake — CHANGES_REQUESTED** [review]

**Comment by masatake** [comment]
@pragmaware, could you look at this one?

**Comment by masatake** [comment]
@drichardson, sorry for being late.
I would like you to change the local variable. Could you fix it and push --force again?
If we cannot get a comment from the parser maintainer, I will merge this.

File: parsers/cxx/cxx_parser_block.c
Comment (drichardson) [reply parsers/cxx/cxx_parser_block.c]: Done

**Comment by drichardson** [comment]
> I would like you to change the local variable. Could you fix it and push --force again?
Done.

**Review by masatake — APPROVED** [review]

**Comment by masatake** [comment]
Thank you.

**Comment by drichardson** [comment]
Time to merge?

**Comment by masatake** [comment]
I will merge this.
@pragmaware, please look at this when you find a time.

**Comment by masatake** [comment]
@drichardson, thank you for the contribution.

**Pull request merged**
